src/components/shared/thread/ActivityGraphFills.js

```diff
@@ -55,75 +82,83 @@ const SMOOTHING_KERNEL: Float32Array = _getSmoothingKernel(
-  treeOrderSampleComparator: ?(
-    IndexIntoSamplesTable,
-    IndexIntoSamplesTable
-  ) => number;
+export class ActivityGraphFillComputer {
+  +lastRenderedGraphSettings: LastRenderedGraphSettings;
+  // The fills and percentages are mutated in place.
+  +mutablePercentageBuffers: SelectedPercentageAtPixelBuffers[];
+  +mutableFills: CategoryFill[];
 
   constructor(
-    canvasPixelWidth: DevicePixels,
-    canvasPixelHeight: DevicePixels,
-    {
-      rangeEnd,
-      rangeStart,
-      categories,
-      interval,
-      samplesSelectedStates,
-      treeOrderSampleComparator,
-      fullThread: { samples, stackTable },
-    }: ActivityGraphProps,
-    categoryDrawStyles: CategoryDrawStyle[]
+    lastRenderedGraphSettings: LastRenderedGraphSettings,
+    mutablePercentageBuffers: SelectedPercentageAtPixelBuffers[],
+    mutableFills: CategoryFill[]
   ) {
-    // Collect the common variables used on the various methods.
-    this.canvasPixelWidth = canvasPixelWidth;
-    this.canvasPixelHeight = canvasPixelHeight;
-    this.rangeEnd = rangeEnd;
-    this.rangeStart = rangeStart;
-    this.rangeLength = rangeEnd - rangeStart;
-    this.categoryDrawStyles = categoryDrawStyles;
-    this.interval = interval;
-    this.xPixelsPerMs = this.canvasPixelWidth / this.rangeLength;
-    this.samples = samples;
-    this.stackTable = stackTable;
-    this.samplesSelectedStates = samplesSelectedStates;
-    this.greyCategoryIndex = categories.findIndex(c => c.color === 'grey') || 0;
-    this.devicePixelRatio = window.devicePixelRatio;
-    this.treeOrderSampleComparator = treeOrderSampleComparator;
-    this.percentageBuffers = _createSelectedPercentageAtPixelBuffers(
-      categoryDrawStyles,
-      this.canvasPixelWidth
-    );
-    this.categoryFills = _getCategoryFills(
-      categoryDrawStyles,
-      this.percentageBuffers
-    );
+    this.lastRenderedGraphSettings = lastRenderedGraphSettings;
+    this.mutablePercentageBuffers = mutablePercentageBuffers;
+    this.mutableFills = mutableFills;
+  }
+
+  /**
+   * Run the computation to compute a list of the fills that need to be drawn for the
+   * ThreadActivityGraph.
+   */
+  run(): void {
+    const {
+      mutableFills,
+      lastRenderedGraphSettings: { canvasPixelWidth },
+    } = this;
+
+    // First go through each sample, and set the buffers that contain the percentage
+    // that a category contributes to a given place in the X axis of the chart.
+    this._accumulateSampleCategories();
+
+    // Smooth the graphs by applying a 1D gaussian blur to the per-pixel
+    // contribution of each fill.
+    for (const fill of mutableFills) {
+      _applyGaussianBlur1D(fill.perPixelContribution, BOX_BLUR_RADII);
+    }
+
+    // Finally compute the last cumulative array.
+    let lastCumulativeArray = mutableFills[0].perPixelContribution;
+    for (const { perPixelContribution } of mutableFills.slice(1)) {
+      for (let i = 0; i < canvasPixelWidth; i++) {
```

Comment:
You could use `perPixelContribution.length` as the upper bound of this loop. Then you don't have to look up the `lastRenderedGraphSettings`.